src/BagReader.js

```diff
@@ -4,82 +4,125 @@
 // found in the LICENSE file in the root directory of this source tree.
 // You may not use this file except in compliance with the License.
 
+// @flow
+
+import { type Callback } from "./types";
+
 import { parseHeader } from "./header";
-import Time from "./Time";
 import nmerge from "./nmerge";
+import { Record, BagHeader, Chunk, ChunkInfo, Connection, IndexData, MessageData } from "./record";
+import { Time } from "./Time";
+
+interface ChunkReadResult {
+  chunk: Chunk;
+  indices: IndexData[];
+}
+
+interface Filelike {
+  read(offset: number, length: number, callback: Callback<Buffer>): void;
+  size(): number;
+}
+
+export type Decompress = {
+  [compression: string]: (buffer: Buffer, size: number) => Buffer,
+};
+
+function cast<T, U: T>(value: T, cls: Class<U> & { name: string }): U {
+  if (!(value instanceof cls)) {
+    throw new Error(`Expected ${cls.name}`);
+  }
+  return value;
+}
+function castArray<T, U: T>(values: T[], cls: Class<U> & { name: string }): U[] {
+  return values.map((x: T) => cast(x, cls));
+}
```

Comment:
Perhaps just loop through the array here & call `cast` on every item rather than mapping to a new array? Since this is pretty high throughput code it might be nice to save a bit of work.

Alternatively what about making `this._readRecordsFromBuffer` both generic & take a `cls` argument as its final argument and it can check the instance of records as it creates them?  Then you could call it like: `const chunkInfos: ChunkInfo[] = this._readRecordsFromBuffer(buffer, count, offset, ChunkInfo)` and you could save an extra iteration just for type checking through the array as well, making the internal method itself a bit more strongly typed.  The internal API would be slightly more verbose, but since it's internal & private that's probably alright.

Reply:
Good suggestion, done!